framework/src/main/java/org/checkerframework/framework/flow/CFAbstractTransfer.java

```diff
@@ -294,10 +294,15 @@ public S initialStore(UnderlyingAST underlyingAST, List<LocalVariableNode> param
         // Allow that local variables are retained; they are effectively final,
         // otherwise Java wouldn't allow access from within the lambda.
         // TODO: what about the other information? Can code further down be simplified?
-        // store.localVariableValues.clear();
         store.classValues.clear();
         store.arrayValues.clear();
-        store.methodValues.clear();
+        // Is there a way to check whether the CF is being run with -AassumePure here?
+        // Best-effort check, ideally we should remove a methodValue if it's impure (not just
+        // non-deterministic)
+        store
+            .methodValues
+            .keySet()
+            .removeIf(methodValue -> !methodValue.isDeterministic(atypeFactory));
```

Comment:
Since we don't know when the code in the lambda is executed, we should assume a side-effecting method has been called. So, I think this should be `methodValues.keySet().removeIf(e -> !e.isUnmodifiableByOtherCode());` which is copied from `CFAbstractStore#updateForMethodCall`.

Reply:
Would that logic hold for the case I've added as a test?
```java
void test(OptContainer container, List<String> strs) {
     if (container.getOpt().isPresent()) {
       strs.forEach(s -> container.getOpt().get()); // Legal
     }
   }
```
Where `getOpt()` is a pure getter. I changed the check to
```java
methodValues.keySet().removeIf(e -> !e.isUnmodifiableByOtherCode());
```
And it fails, is this a false positive we can't do anything about?

I did add an additional condition

```java
methodValues.keySet().removeIf(e -> !e.isUnmodifiableByOtherCode() && !ElementUtils.isGetter(methodValue.element));
```
Which makes the new test case pass.